addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SinkDetectionVerifyProbableSinks.java

```diff
@@ -0,0 +1,142 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrules;
+
+import static org.zaproxy.zap.extension.ascanrules.SinkDetectionCollectAndRefreshParamValues.SINK_DETECTION_STORAGE;
+
+import java.util.List;
+import java.util.Random;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.AbstractAppParamPlugin;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.core.scanner.Category;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.addon.commonlib.ParamSinksUtils;
+
+public class SinkDetectionVerifyProbableSinks extends AbstractAppParamPlugin {
+
+    /** Prefix for internationalised messages used by this rule */
+    private static final String MESSAGE_PREFIX = "ascanrules.sinkdetectionverifyprobablesinks.";
+
+    private static Logger log = Logger.getLogger(SinkDetectionVerifyProbableSinks.class);
+
+    @Override
+    public int getId() {
+        return 40037;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public int getCategory() {
+        return Category.INJECTION;
+    }
+
+    @Override
+    public String[] getDependency() {
+        return new String[] {"SinkDetectionFindProbableSinks"};
+    }
+
+    @Override
+    public String getSolution() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public String getReference() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public void scan(HttpMessage msg, String param, String value) {
+        // TODO: should test only on POST/PUT ?
+        SinkDetectionStorage storage = getStorage();
+        List<HttpMessage> possibleSinks = storage.getPossibleSinksForValue(value);
+
+        try {
+            HttpMessage inputMsg = msg.cloneRequest();
+            String injectedValue = getRandomString();
+            this.setParameter(inputMsg, param, injectedValue);
+            sendAndReceive(inputMsg, false);
+
+            for (HttpMessage possibleSink : possibleSinks) {
+                // Reflections on the same request are not considered "stored sinks"
+                if (inputMsg.hashCode() != possibleSink.hashCode()) {
+                    HttpMessage possibleSink1 = possibleSink.cloneRequest();
```

Comment:
can be named something like possibleSinkClone or clonePossibleSink so something else.

Reply:
Yes that's better than possibleSink1, I will change it